### PR TITLE
[Narwhal] a few cleanup

### PR DESCRIPTION
**Pull request by mwtian**
1. If a certificate is written to the local store but not sent to the DAG, there will be inconsistencies in the DAG if Narwhal continues to operate. A certificate may assume its parents exist in the DAG but actually not. So add a warning when this happens.
2. Logic to check existence of certificate parents can be simplified, since Narwhal no longer updates committee and epochs dynamically.

### CODE REVIEW

File: narwhal/storage/src/certificate_store.rs [hunk 1]
```diff
@@ -177,6 +177,18 @@ impl CertificateStore {
         }
     }
 
+    /// Retrieves a certificate from the store. If not found
+    /// then None is returned as result.
+    pub fn contains_digest(&self, id: &CertificateDigest) -> StoreResult<bool> {
```
Comment (akichidis) [inline narwhal/storage/src/certificate_store.rs:182]: I would just go with `contains`

File: narwhal/primary/src/core.rs [hunk 2]
```diff
@@ -608,19 +610,30 @@ impl Core {
...
+        }
 
         // Send it to the consensus layer.
-        let digest = certificate.header.digest();
         if let Err(e) = self.tx_new_certificates.send(certificate).await {
             warn!(
                 "Failed to deliver certificate {} to the consensus: {}",
                 digest, e
             );
+            return Err(DagError::ShuttingDown);
```
Comment (akichidis) [inline narwhal/primary/src/core.rs:634]: good point!

File: narwhal/primary/src/core.rs [hunk 3]
```diff
@@ -608,19 +610,30 @@ impl Core {
...
+        if let Err(e) = self
+            .append_certificate_in_aggregator(certificate.clone())
+            .await
+        {
+            warn!(
+                "Failed to aggregate certificate {} for header: {}",
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
```
Comment (akichidis) [inline narwhal/primary/src/core.rs:626]: If you don't want to repeat the `DagError::ShuttingDown` you can instead use the `tap_err` method to use the returned `Result` and do:

```suggestion
        self.append_certificate_in_aggregator(certificate.clone())
            .await
            .tap_err(|err| {
                warn!(
                    "Failed to aggregate certificate {} for header: {}",
                    digest, err
                )
            })?;
```

**Review by akichidis — APPROVED** [review]
Thanks a lot @mwtian !

File: narwhal/primary/src/core.rs (hunk 2 above)
Comment (mwtian) [reply narwhal/primary/src/core.rs:634]: I hope later on we can verify the system is indeed shutting down.

File: narwhal/primary/src/core.rs (hunk 3 above)
Comment (mwtian) [reply narwhal/primary/src/core.rs:626]: Good point, this is a good alternative. But I do want to make sure only `DagError::ShuttingDown` is returned from this point. Although this is indeed the only possible error from `append_certificate_in_aggregator()`, it is not obvious.

**Pull request merged**